modules/hisilicon/hisi_play.c

```
@@ -0,0 +1,157 @@
+/**
+ * @file hisi_play.c  HiSilicon sound driver - player
+ *
+ * Copyright (C) 2022 Dmitry Ilyin
+ */
+
+#include <re.h>
+#include <rem.h>
+#include <baresip.h>
+#include <threads.h>
+
+#include "hisi.h"
+
+#include <mpi_audio.h>
+#include <mpi_sys.h>
+
+struct auplay_st {
+	thrd_t thread;
+	volatile bool run;
+	void *sampv;
+	size_t sampc;
+	auplay_write_h *wh;
+	void *arg;
+	struct auplay_prm prm;
+};
+
+
+static void auplay_destructor(void *arg)
+{
+	struct auplay_st *st = arg;
+
+	/* Wait for termination of other thread */
+	if (st->run) {
+		debug("hisi: stopping playback thread\n");
+		st->run = false;
+		thrd_join(st->thread, NULL);
+	}
+
+	int ret = HI_MPI_AO_DisableChn(0, 0);
+	if (HI_SUCCESS != ret) {
+		printf("error %d\n", ret);
```

consider ignoring errors on close, or use warning() instead